c-s runtime: write and read scenarios

muzarski opened the pull request:
# Changes
- Cosmetic changes in the settings module that simplify the access to the parameters (from the main).
- Introduced `operation` module.
- Implemented row-generation logic, based on c-s logic.
- Implemented write and read operations.
- Implemented main logic responsible for running the actual runtime scenario.

piodul commented on src/bin/cql-stress-cassandra-stress/settings/command/help.rs:
```diff
@@ -13,24 +13,24 @@ pub fn print_help() {
         Ok(cmd) => {
             cmd.print_help();
-            Ok(CommandParams::Special)
+            Ok(None)
```
> This commit introduces very subtle changes in the settings module.
> It helps the users of the settings access the configuration in a more
> convenient way.

I'd appreciate if the first part of the commit message were less vague. A commit should concentrate on doing one thing, and "very subtle changes in the settings module" is very permissive with regard to what can happen there.

It looks like the idea is to replace `CommandParams` with `Option<CommandParams>` and get rid of `CommandParams::Special`, right? I think you should mention it. The justification that it's more convenient (idiomatic?) is already good.

Besides, a more specific description slightly helps me because I don't have to spend as much time trying to _guess_ the exact idea behind the commit, instead I can concentrate on _verifying_ whether your code aligns with your intentions.

piodul commented on src/bin/cql-stress-cassandra-stress/settings/command/help.rs:
```diff
@@ -13,24 +13,24 @@ pub fn print_help() {
 
-fn print_option_help(option_str: &str) -> Result<CommandParams> {
+fn print_option_help(option_str: &str) -> Result<Option<CommandParams>> {
     Options::print_help(option_str)?;
-    Ok(CommandParams::Special)
+    Ok(None)
 }
```
I'm not sure if I understand why this function returned `Result<CommandParams>` in the first place. It is supposed to only print help, right? Returning `Result<>` I can understand, but `CommandParams`?

piodul commented on src/bin/cql-stress-cassandra-stress/settings/option/rate.rs:
```diff
@@ -5,7 +5,7 @@ use crate::settings::{
 
 pub struct RateOption {
-    threads_info: ThreadsInfo,
+    pub threads_info: ThreadsInfo,
```
Is this change supposed to be included in this commit?

muzarski replied:
Extracted to yet another commit

piodul commented on src/bin/cql-stress-cassandra-stress/settings/mod.rs:
```diff
@@ -22,7 +23,7 @@ use self::option::SchemaOption;
 pub struct CassandraStressSettings {
     pub command: Command,
-    pub params: CommandParams,
+    pub command_params: CommandParams,
```
This rename should probably be a separate commit. It looks like you really are mixing multiple things in the first commit, but you shouldn't.

piodul commented on src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:
```diff
@@ -327,11 +327,16 @@ fn parse_with_handles(handles: ReadWriteParamHandles) -> ReadWriteParams {
 
-pub fn parse_read_write_params(cmd: &Command, payload: &mut ParsePayload) -> Result<CommandParams> {
+pub fn parse_read_write_params(
+    cmd: &Command,
+    payload: &mut ParsePayload,
+) -> Result<Option<CommandParams>> {
     let args = payload.remove(cmd.show()).unwrap();
     let (parser, handles) = prepare_parser(cmd.show());
     parser.parse(args)?;
-    Ok(CommandParams::BasicParams(parse_with_handles(handles)))
+    Ok(Some(CommandParams {
+        basic_params: parse_with_handles(handles),
+    }))
```
This function always returns either `Err()` or `Ok(Some(...))`. Does it make sense for it to return `Result<Option<CommandParams>>` instead of `Result<CommandParams>`?

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/values/blob.rs:
```diff
@@ -1,5 +1,7 @@
 use std::cmp::min;
 
+use scylla::frame::response::result::CqlValue;
```
> c-s values_generator: return CqlValue

Nit: typo in the commit message: `implementiation` -> `implementation`

piodul commented on src/bin/cql-stress-cassandra-stress/operation/row_generator.rs:
```diff
@@ -0,0 +1,147 @@
+/// I think it's a great place to address how read and write workloads cooperate.
+/// For reference, see: https://github.com/scylladb/cql-stress/pull/43#discussion_r1304274035.
+///
+/// At first, we need to notice that as long as `pk_seed_distribution` is a DETERMINISTIC distribution,
+/// the row-generation logic is also deterministic.
```
What is a deterministic distribution? Is this concept defined somewhere?

muzarski replied:
Added description in the comment.

piodul commented on src/bin/cql-stress-cassandra-stress/operation/row_generator.rs:
```diff
@@ -0,0 +1,147 @@
+/// To be more precise: any read workload that samples the partition_key seeds from the distribution
+/// sampling the values from the subset of range 1..100 (one used in the write routine) will successfully validate the data.
+///
+/// This also means, that if we used a NON-DETERMINISTIC distribution e.g. `UniformDistribution` (which is by default seeded with current time as millis)
+/// in the write workload, most of the times, the read workload will result in a validation error.
+/// For example:
```
Side note: it's technically possible to achieve determinism with non-seq distributions, and AFAIR we already do this in the s-b command - you can use the operation number as the seed to the random generator, and then use it to generate the partition key. I'm not sure why the original c-s doesn't do it...

Perhaps it's worth considering this as a future improvement.

muzarski replied:
Good point, and to be quite honest, I'd really love to follow the approach you mentioned. I'm not sure what c-s wanted to achieve by being non-deterministic in the first place.

I'll save this conversation for the future (in the comment), and hopefully we will discuss it with @avelanarius .

muzarski commented on src/bin/cql-stress-cassandra-stress/java_generate/values/blob.rs:
```diff
@@ -43,12 +43,17 @@ mod tests {
     /// Utility function that maps u8 vector values to i8 values.
-    fn to_vec_i8(v: Vec<u8>) -> Vec<i8> {
-        v.into_iter().map(|x| x as i8).collect::<Vec<i8>>()
+    fn to_vec_i8(v: CqlValue) -> Vec<i8> {
+        v.as_blob()
+            .unwrap()
+            .iter()
+            .map(|x| *x as i8)
+            .collect::<Vec<i8>>()
```
It's a test-utility function. There were no changes to the test itself, that's why we don't see the usage in this commit.

piodul replied:
Ah, OK. For some reason I thought that this is a new function, but indeed it is not.

muzarski commented on src/bin/cql-stress-cassandra-stress/settings/command/mod.rs:
```diff
@@ -77,31 +77,24 @@ impl Command {
-
-    // USER
-    // TODO: UserParams,
+pub struct CommandParams {
+    // Parameters shared across all of the commands
+    pub basic_params: ReadWriteParams,
+    // TODO:
+    // mixed_params: Option<MixedParams>
+    // user_params: Option<UserParams>
```
If you look at the parameters of some commands, you can see that parameters of `read/write` are a subset of parameters of both `mixed` and `user` command. IMO, it makes sense to get rid of the enum, and represent it as struct, since both `mixed` and `user` will include `basic_params`, and some additional parameters specific to them.

piodul replied:
Ok. Again, this should be mentioned in the commit message.

muzarski commented:
v2:
Fixed review comments, including (the most important):
- split the first commit into 3 commits
- extended commit message regarding `CommandParams` refactor

muzarski commented:
v3:
Rebased on top of master

piodul commented on src/bin/cql-stress-cassandra-stress/operation/mod.rs:
```diff
@@ -21,3 +23,41 @@ fn recompute_seed(seed: i64, partition_key: &CqlValue) -> i64 {
+
+    anyhow::ensure!(
+        first_row.columns.len() == generated_row.len(),
+        "Generated row's length: {}, result row's length: {}",
```
How about printing those rows?

piodul commented on src/bin/cql-stress-cassandra-stress/operation/mod.rs:
```diff
@@ -21,3 +23,41 @@ fn recompute_seed(seed: i64, partition_key: &CqlValue) -> i64 {
+                None => false,
+            });
+
+    anyhow::ensure!(result, "The data doesn't match.");
```
How about printing those rows so that the user can see how the mismatch looks like?

piodul reviewed (commented):
Some nitpicks related to error messages, but otherwise LGTM
